### PR TITLE
cgroup: add PIDs cgroup controller support

### PR DESCRIPTION
**Pull request by cyphar**
This is a fixed up version of the now-reverted #58. The main issue with that code was that **some of** the memory cgroup's limits weren't being set at all if you were using systemd. Unfortunately, I seem to be unable to test the systemd portion (and the systemd test cases for memory don't actually test that the limits cause a process to fail).

All other cgroups were unaffected by this bug (as their limits were set using `.Set()`, and were joined either explicitly or using systemd), regardless of whether they were unsupported, partial supported or fully supported by systemd.

Some testing has revealed that the old setup didn't work either (because `getSubsystemPath` doesn't appear to work as expected).

/cc @mrunalp @hqhq @vishh @crosbymichael @dqminh

### CODE REVIEW

**Comment by cyphar** [comment]
@mrunalp The main change I made to #58 was to actually explicitly `join` the cgroups and set memory. In addition, there we a couple of subtle bugs (needed to move the `CheckCpushares` check). After building Docker with my patches, systemd support appears to work (it doesn't work on `master` right now).

**Comment by cyphar** [comment]
/ping @mrunalp @hqhq @vishh @crosbymichael @dqminh @LK4D4

**Comment by hqhq** [comment]
What problems are you saying about systemd cgroup in docker upstream exactly? I'm sure we have some systemd cgroup users and we don't see reports from them. And what's the problem about `getSubsystemPath` in systemd cgroup? Can you elaborate?

**Comment by cyphar** [comment]
@hqhq Well, if you try to use the version on master and set something like `--cpu-shares` you get errors when the container tries to start. There's also a few issues we've seen with customers where systemd randomly starts moving processes around slices.

**Comment by hqhq** [comment]
@cyphar No I don't get any errors when setting `--cpu-shares` with the latest docker version on master, my box is centos 7.1 with systemd-219, but I do know there might be problems with some very new systemd versions, what systemd version are you using.

BTW, I don't have a box with latest kernel to test this PR out, but over all the codes in this PR looks good, if there are any systemd issues with docker, I can help diagnose.

**Comment by cyphar** [comment]
@hqhq I'm testing this on Arch (with a custom-compiled kernel). The systemd problems I'm referring to are not related to this PR (there are a few). I'll open a separate issue (or PR) for these problems if appropriate (a SUSE customer has seen issues with systemd on SLE).

**Comment by cyphar** [comment]
PTAL /cc @mrunalp @hqhq @vishh @crosbymichael @dqminh

**Comment by mrunalp** [comment]
Do you have a docker branch with this integrated?

**Comment by LK4D4** [comment]
@mrunalp docker/docker#18697

**Comment by LK4D4** [comment]
I tried with docker. It works perfectly with fs cgroups. But seems like systemd cgroups doesn't work at all in master, so I can't test it.

**Comment by mrunalp** [comment]
@LK4D4 Here is a fix for systemd in docker https://github.com/docker/docker/pull/19149

**Comment by cyphar** [comment]
@LK4D4 @mrunalp note that you'll have to re-run `hack/vendor.sh`, since @jfrazelle hasn't updated her branch yet.

**Comment by LK4D4** [comment]
I tested definitely with this branch :)

**Comment by cyphar** [comment]
Good to hear. :smile_cat:

**Comment by LK4D4** [comment]
I tested and it works with systemd.

**Comment by LK4D4** [comment]
LGTM

**Comment by cyphar** [comment]
Due to some upstream kernel changes introduced in `v4.4` (namely, allowing cgroups to track zombies) the tests that assume a small `pids.max` will work are broken. This is not because of us (we're applying as late as possible), it's because the kernel keeps the counter up for long enough for us to hit it. I'll update the tests to reflect this.

**Comment by cyphar** [comment]
In any case, PTAL so we can merge this soon.

/ping @mrunalp @docker/maintainers

**Comment by mrunalp** [comment]
@cyphar Yes, I am going to test this today. Thanks!

File: libcontainer/integration/exec_test.go [hunk 1]
```diff
@@ -525,6 +525,75 @@ func testCpuShares(t *testing.T, systemd bool) {
...
+	if err != nil && strings.Contains(err.Error(), "no such directory for pids.max") {
+		t.Skip("PIDs cgroup is unsupported")
+	}
+	ok(t, err)
+
+	if ret != 0 {
+		t.Fatalf("expected fork() to succeed with permissive pids limit")
+	}
+
+	// Enforce a restrictive limit limit (shell + 6 * true + 3).
```
Comment (mrunalp) [inline libcontainer/integration/exec_test.go]: s/limit limit/limit/

File: libcontainer/process_linux.go [hunk 2]
```diff
@@ -229,13 +231,58 @@ func (p *initProcess) start() (err error) {
...
+			if err := json.NewDecoder(p.parentPipe).Decode(&ierr); err != nil && err != io.EOF {
+				return newSystemError(err)
+			}
+			if ierr != nil {
+				break loop
+			}
+			// Programmer error.
+			panic("No error following JSON procError payload.")
+		default:
+			return newSystemError(fmt.Errorf("Invalid JSON synchronisation payload from child."))
```
Comment (mrunalp) [inline libcontainer/process_linux.go]: Start error string with lower case.

Comment (mrunalp) [inline libcontainer/process_linux.go]: Also, no punctuation. We are following this convention https://github.com/golang/go/wiki/CodeReviewComments#error-strings

File: libcontainer/process_linux.go [hunk 3]
```diff
@@ -229,13 +231,58 @@ func (p *initProcess) start() (err error) {
...
+		}
+	}
+	if !sentRun {
+		return newSystemError(fmt.Errorf("Couldn't synchronise with container process."))
```
Comment (mrunalp) [inline libcontainer/process_linux.go]: Same here.

**Comment by mrunalp** [comment]
I tested and it worked fine for me. Besides the nits, LGTM

**Comment by cyphar** [comment]
@mrunalp Nits addressed.

**Comment by mrunalp** [comment]
@cyphar Thanks! LGTM

**Comment by yijia2413** [comment]
great!!

**Pull request merged**